client/allocrunner/taskrunner/validate_hook.go

```diff
@@ -48,10 +48,21 @@ func validateTask(task *structs.Task, taskEnv *taskenv.TaskEnv, conf *config.Con
 		[]string{"user.denylist", "user.blacklist"},
 		config.DefaultUserDenylist,
 	)
+
+	allowedUsers := conf.ReadStringListAlternativeToMapDefault(
+		[]string{"user.allowlist", "user.whitelist"},
+		config.DefaultUserAllowlist,
+	)
+
+	// if you are on the allowlist, you are allowed
+	// if you are not on the allowlist, check the denylist
+	// if you are on the denylist, you are disallowed
 	checkDrivers := conf.ReadStringListToMapDefault("user.checked_drivers", config.DefaultUserCheckedDrivers)
 	if _, driverMatch := checkDrivers[task.Driver]; driverMatch {
-		if _, unallowed := unallowedUsers[task.User]; unallowed {
-			mErr.Errors = append(mErr.Errors, fmt.Errorf("running as user %q is disallowed", task.User))
+		if _, allowed := allowedUsers[task.User]; !allowed {
+			if _, unallowed := unallowedUsers[task.User]; unallowed {
+				mErr.Errors = append(mErr.Errors, fmt.Errorf("running as user %q is disallowed", task.User))
+			}
 		}
 	}
```

Comment:
The logic here currently "fails open". Consider the following configuration:

```hcl
client {
  options = {
    "user.allowlist" = "nobody"
  }
}
```

If I try to run a job with `user = "root"`, we'd check the allowlist and not find them, but not find them on the denylist either, so they'd be allowed to go ahead. Whereas an allowlist should _only_ allow those users on the allowlist and not on the denylist.

Likewise, consider the following misconfiguration:

```hcl
client {
  options = {
    "user.allowlist" = "root"
    "user.denylist" = "root,ubuntu"
  }
}
```

In this case, `root` would be allowed even though under a "default deny" approach the denylist should override the allowlist.